# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/PersistentBufferProvider.h
# Commit: 7b82837e2f56
# Full Hash: 7b82837e2f56582e1e86897285f4109856b72000
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2021-11-11 15:25:44
# Regressor Bug: 1739448
# File Overlap Count: 1
# Description:
#   Bug 1739448 - Implement a prototype WebGL-backed Canvas2D. r=gfx-reviewers,nical,jrmuizel
#   
#   This mainly provides DrawTargetWebgl, which implements the subset of the DrawTarget
#   API necessary for integration with CanvasRenderingContext2D. It translates them to
#   suitable commands for its internal ClientWebGLContext, which then manages remoting
# ==============================================================================

diff -r feac99f1dae3 -r 7b82837e2f56 gfx/layers/PersistentBufferProvider.h
--- a/gfx/layers/PersistentBufferProvider.h	Thu Nov 11 04:33:50 2021 +0000
+++ b/gfx/layers/PersistentBufferProvider.h	Thu Nov 11 04:33:50 2021 +0000
@@ -18,6 +18,8 @@
 
 namespace mozilla {
 
+class ClientWebGLContext;
+
 namespace gfx {
 class SourceSurface;
 class DrawTarget;
@@ -41,7 +43,8 @@
 
   virtual ~PersistentBufferProvider() = default;
 
-  virtual LayersBackend GetType() { return LayersBackend::LAYERS_NONE; }
+  virtual bool IsShared() const { return false; }
+  virtual bool IsAccelerated() const { return false; }
 
   /**
    * Get a DrawTarget from the PersistentBufferProvider.
@@ -67,6 +70,8 @@
 
   virtual TextureClient* GetTextureClient() { return nullptr; }
 
+  virtual ClientWebGLContext* AsWebgl() { return nullptr; }
+
   virtual void OnShutdown() {}
 
   virtual bool SetKnowsCompositor(KnowsCompositor* aKnowsCompositor) {
@@ -96,8 +101,6 @@
 
   explicit PersistentBufferProviderBasic(gfx::DrawTarget* aTarget);
 
-  LayersBackend GetType() override { return LayersBackend::LAYERS_BASIC; }
-
   already_AddRefed<gfx::DrawTarget> BorrowDrawTarget(
       const gfx::IntRect& aPersistedRect) override;
 
@@ -114,13 +117,28 @@
  protected:
   void Destroy();
 
- private:
-  virtual ~PersistentBufferProviderBasic();
+  ~PersistentBufferProviderBasic() override;
 
   RefPtr<gfx::DrawTarget> mDrawTarget;
   RefPtr<gfx::SourceSurface> mSnapshot;
 };
 
+class PersistentBufferProviderAccelerated
+    : public PersistentBufferProviderBasic {
+ public:
+  MOZ_DECLARE_REFCOUNTED_VIRTUAL_TYPENAME(PersistentBufferProviderAccelerated,
+                                          override)
+
+  explicit PersistentBufferProviderAccelerated(gfx::DrawTarget* aTarget);
+
+  bool IsAccelerated() const override { return true; }
+
+  ClientWebGLContext* AsWebgl() override;
+
+ protected:
+  ~PersistentBufferProviderAccelerated() override;
+};
+
 /**
  * Provides access to a buffer which can be sent to the compositor without
  * requiring a copy.
@@ -135,7 +153,7 @@
       gfx::IntSize aSize, gfx::SurfaceFormat aFormat,
       KnowsCompositor* aKnowsCompositor);
 
-  LayersBackend GetType() override;
+  bool IsShared() const override { return true; }
 
   already_AddRefed<gfx::DrawTarget> BorrowDrawTarget(
       const gfx::IntRect& aPersistedRect) override;